feat: consistent handling of edge cases for header sync

hansieodendaal opened the pull request:
Description
---
This PR fixes an edge case for header sync where the local node has more headers, but the remote node has better proof-of-work (accumulated difficulty), thus higher actual blockchain height. It also simplifies the attempted header sync logic and error handling when a peer does not return headers and introduces consistent naming, variable re-use and use of information

Supersedes #5756

Motivation and Context
---
When doing header sync, and determining from which peer to sync, we need to consider our actual blockchain state when comparing chains. It is possible that our node has valid block headers that will translate into a  higher proof of work when fully synced than the remote syncing peer, but lacking the blocks to back the block headers, thus the remote chain has an actual higher proof of work blockchain.
With the current implementation, race conditions can exist when determining if the peer is misbehaving, i.e. lying about their proof-of-work or not wanting to supply block headers, due to a mismatch in tip height used for the check.

This PR fixes the race conditions by always using the latest local chain metadata and block headers, ignoring all sync peers that do not exceed our own accumulated difficulty and using consistent information in all the checks.

How Has This Been Tested?
---
Added integration-level unit tests:
- `test_header_sync_happy_path`
- `test_header_sync_with_fork_happy_path`
- `test_header_sync_uneven_headers_and_blocks_happy_path`
- `test_header_sync_uneven_headers_and_blocks_peer_lies_about_pow_no_ban`
- `test_header_sync_even_headers_and_blocks_peer_lies_about_pow_with_ban`
- `test_header_sync_even_headers_and_blocks_peer_metadata_improve_with_reorg`

What process can a PR reviewer use to test or verify this change?
---
- Code walk-through
- Review unit tests

<!-- Checklist -->
<!-- 1. Is the title of your PR in the form that would make nice release notes? The title, excluding the conventional commit
tag, will be included exactly as is in the CHANGELOG, so please think about it carefully. -->


Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

<!-- Does this include a breaking change? If so, include this line as a footer -->
<!-- BREAKING CHANGE: Description what the user should do, e.g. delete a database, resync the chain -->

SWvheerden committed:
Add consistency to header sync

This PR fixes an edge case for header sync where the local node has a higher chain
header, but the remote node has a higher actual blockchain height. It also simplifies
the attempted header sync logic and error handling when a peer does not return headers
and introduces consistent naming, variable re-use and use of information.

When doing header sync, and determining from which peer to sync, we need to consider
our actual blockchain state when comparing chains. It is possible that our node has
valid block headers that will translate into a higher proof of work when fully synced
than the remote syncing peer, but lacking the blocks to back the block headers, thus
the remote chain has an actual higher proof of work blockchain.

With the current implementation, race conditions can exist when determining if the
peer is misbehaving, i.e. lying about their proof-of-work or not wanting to supply
block headers, due to a mismatch in tip height used for the check.

This PR fixes the race conditions by always using the latest local chain metadata
and block headers, ignoring all sync peers that do not exceed our own accumulated
difficulty and using consistent information in all the checks.

Added a header sync tests integration unit test.

hansieodendaal commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
             .collect::<Result<Vec<_>, _>>()
             .map_err(BlockHeaderSyncError::ReceivedInvalidHeader)?;
-        let num_new_headers = headers.len();
+        let num_new_headers = headers.len(); // Required fro later use, no 'Copy' trait on 'BlockHeader'
```
```suggestion
        let num_new_headers = headers.len(); // Required for later use, no 'Copy' trait on 'BlockHeader'
```

sdbondi reviewed (dismissed):
Nice - utACK

SWvheerden commented on applications/minotari_console_wallet/src/ui/components/transactions_tab.rs:
```diff
@@ -388,21 +388,23 @@ impl TransactionsTab {
 
-            // let mined_time = DateTime::<Local>::from_utc(tx.mined_timestamp, Local::now().offset().to_owned());
+            // let mined_time = DateTime::<Local>::from_naive_utc_and_offset(tx.mined_timestamp,
+            // Local::now().offset().to_owned());
```
```suggestion
```

SWvheerden commented on applications/minotari_console_wallet/src/ui/components/transactions_tab.rs:
```diff
@@ -388,21 +388,23 @@ impl TransactionsTab {
                             .format("%Y-%m-%d %H:%M:%S")
                     ),
                 },
                 // format!("{}", mined_time.format("%Y-%m-%d %H:%M:%S")),
```
```suggestion
```

SWvheerden commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
                 debug!(
                     target: LOG_TARGET,
-                    "Peer `{}` has sent no headers but forked_hash_index is {}. The peer is behind our chain.",
+                    "Peer `{}` has sent no headers with forked_hash_index {}. The peer has less headers than we have.",
```
```suggestion
                    "Peer `{}` has sent no headers with some forked_hash_index {}. The peer has less headers than we have.",
```

SWvheerden commented on base_layer/core/tests/tests/header_sync.rs:
```diff
@@ -0,0 +1,294 @@
+        StateEvent::HeadersSynchronized(val, sync_result) => {
+            assert_eq!(val.claimed_chain_metadata().height_of_longest_chain(), 1);
+            assert_eq!(val.claimed_chain_metadata().accumulated_difficulty(), 4);
+            assert_eq!(sync_result.headers_returned, 1);
```
Nit: This is probably fine as is here. I would have preferred to test these values in unit tests, then we don't have to add code to allow testing of values. These values are only used by these tests.

hansieodendaal replied:
Agreed that it is probably fine

SWvheerden reviewed (commented):
looks good, just needs a few commented out lines removed

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -245,59 +249,69 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
+            ));
+        }
+
+        let peer_response = self
```
What is the purpose of moving this logic here? I would prefer to keep the logic in `determine_sync_status` but propagate `peer_response` as part of the Error or result

hansieodendaal replied:
It was for testability. Moving the `find_chain_split` back to `determine_sync_status` and returning `peer_response` alongside `header_sync_status`, as `AttemptSyncResult` pass back `header_sync_status` and that will become bloated.

SWvheerden replied:
We can move the call back in, but move the decision logic out. 

```rust
fn determine_sync_status(...)->{
 let resp = self. find_chain_split(..)
 decide_logic(resp)
}

fn decide_logic(..)->
{
..
}
```

Something like that.

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -245,59 +249,69 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
+            .await?;
+
+        match header_sync_status.clone() {
+            HeaderSyncStatus::InSyncOrAhead => {
```
```suggestion
            HeaderSyncStatus::InSyncOrAhead{ peer_response } => {
```
Rather propagate it here

hansieodendaal replied:
As above

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -245,59 +249,69 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
-                         {})",
-                        header_tip_height,
-                        metadata.height_of_longest_chain(),
+                         {}), peer has lied about chain metadata or did not want to provide headers",
```
This is not true. We only checked if our best_block < our best_headers

SWvheerden replied:
Its true by deduction. 
For us to have entered Header_sync the peer needed tot have a better pow(it claimed it did)
Then we check that out pow did not update and that they still have a claimed higher pow. 

We then asked the peer to provide us the chain which had a better pow, but it could not. So we have the same or more headers on the exact same chain as they have.  But we can have the headers, not the block, so if we check if we have the same blocks than headers, we must have the same or more blocks than they. Its not possible to have blocks and not the headers. 

So the only conclusion we can make is one of the following 3:
They lied about their pow
They did not want to provide us with a higher chain
The swapped to a lower-difficulty pow chain. 

So we are only left with 1 and 2 realisticly.

hansieodendaal replied:
Ok, I provided better data for a better decision

stringhandler replied:
Please provide a link to the new code or more detail about what you did, I can't find it among the other changes

hansieodendaal replied:
The conditions are now tested separately and both will be ban-able offences.
See
```rust
        // Basic sanity check about the peer's claimed chain metadata
        if chain_split_result.peer_accumulated_difficulty < sync_peer.claimed_chain_metadata().accumulated_difficulty()
```
and
```rust
        match header_sync_status.clone() {
            HeaderSyncStatus::InSyncOrAhead => {
                // Our POW is less than the peer's POW, as verified before the attempted header sync, therefore, if the
                // peer did not supply any headers and we know we are behind based on the peer's claimed metadata, then
                // we can ban the peer.
                if best_header.height() == best_block.height() {
```

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -245,59 +249,69 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
                     );
+
                     Err(BlockHeaderSyncError::PeerSentInaccurateChainMetadata {
                         claimed: sync_peer.claimed_chain_metadata().accumulated_difficulty(),
```
This seems like an assumption, the logic here is incorrect. See above comment

hansieodendaal replied:
As per above

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -245,59 +249,69 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
                 self.hooks.call_on_progress_header_hooks(
                     split_info
-                        .local_tip_header
+                        .best_block
```
Should this not be `best_header`?

hansieodendaal replied:
I provided a better name

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
-        if headers.is_empty() {
-            if fork_hash_index > 0 {
+        // If the peer returned no new headers, they have no headers, but may still have more blocks than we have,
+        // thus have a higher accumulated difficulty.
```
This comment makes no sense. The headers are what determines the accumulated difficulty.

SWvheerden replied:
This comment exactly states the problem we need to address, the logic is divided into weird places. 

The reason this is, is that the PoW of the local node is determined by the headers, of the blocks we have the complete block for. 
But the node downloads the headers before it downloads the complete block. 

So in this case, the node could already have downloaded the headers, but not yet the blocks, so technically the pow of the local node is lower, although it might have the headers already.

hansieodendaal replied:
Fixed

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
                 );
-
-                return Ok(SyncStatus::WereAhead);
+            } else {
```
This feels like the peer can abuse this. They return no headers but then return a fork index of 0.

How is it that we requested headers from them but didn't receive any? 

Feels like we should ban them

SWvheerden replied:
See the above comment, we first need to check if they have full blocks for those, as we might not have them yet. 
But if we have the full blocks for those, then yes, we do ban them, but only after that check

hansieodendaal replied:
See new logic test ^^

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
             .collect::<Result<Vec<_>, _>>()
             .map_err(BlockHeaderSyncError::ReceivedInvalidHeader)?;
-        let num_new_headers = headers.len();
+        let num_new_headers = headers.len(); // Required for later use, no 'Copy' trait on 'BlockHeader'
```
```suggestion
        let num_new_headers = headers.len(); 
```

No need to explain

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
-        let chain_split_hash = block_hashes.get(fork_hash_index as usize).unwrap();
+        let chain_split_hash = peer_response
+            .block_hashes
+            .get(peer_response.fork_hash_index as usize)
```
This unwrap seems like the peer could cause a panic if `fork_hash_index > block_hashes.len()`

SWvheerden replied:
That is not possible, it is checked above, with a comment git is not showing
https://github.com/tari-project/tari/blob/6e52f447a5f127af4a7523fb0f26684bc9e9173b/base_layer/core/src/base_node/sync/header_sync/synchronizer.rs#L442C11-L443C79

hansieodendaal replied:
Simplified - `find_chain_split` will return the `chain_split_hash `

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -462,21 +487,21 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
-        if remote_tip_height < split_height {
+        let split_height = best_header.height().saturating_sub(peer_response.reorg_steps_back);
+        if peer_response.remote_tip_height < split_height {
             return Err(BlockHeaderSyncError::InvalidProtocolResponse(format!(
```
Does this result in a ban? I think it should

hansieodendaal replied:
Yes it does

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -462,21 +487,21 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
-            local_tip_header,
-            remote_tip_height,
-            reorg_steps_back: steps_back,
+            best_block,
```
did you mean `best_header`?

hansieodendaal replied:
Fixed the name

stringhandler commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -781,18 +797,43 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
+#[derive(Debug, Clone, PartialEq)]
+pub struct ChainSplitInfo {
+    /// The best block on the local chain.
+    pub best_block: ChainHeader,
```
This is badly named. Is it the best header we have or the best block we have?

I made a few comments already, but looks like it could just be that this should have been `best_header`

SWvheerden replied:
This is exactly the fix we made, its the header of the best block. 
Best block and best header are not always the same

hansieodendaal replied:
Fixed the name

stringhandler commented on base_layer/core/tests/helpers/nodes.rs:
```diff
@@ -105,6 +108,7 @@ pub struct BaseNodeBuilder {
     mempool_config: Option<MempoolConfig>,
     mempool_service_config: Option<MempoolServiceConfig>,
     liveness_service_config: Option<LivenessConfig>,
+    p2p_config: Option<P2pConfig>,
```
Not related to this PR, but this struct should really be called "TestBaseNodeBuilder"

stringhandler commented on base_layer/p2p/src/initialization.rs:
```diff
@@ -135,7 +135,7 @@ pub async fn initialize_local_test_comms<P: AsRef<Path>>(
     seed_peers: Vec<Peer>,
     shutdown_signal: ShutdownSignal,
-) -> Result<(CommsNode, Dht, MessagingEventSender), CommsInitializationError> {
+) -> Result<(UnspawnedCommsNode, Dht, MessagingEventSender), CommsInitializationError> {
```
Unrelated: This test method should also be moved into a file that shows it's a test method. Because the method is so long, it is difficult to see that it's test code

Commit:
review comments

Commit:
added more unit tests

SWvheerden commented on base_layer/core/src/base_node/proto/rpc.proto:
```diff
@@ -51,6 +51,8 @@ message FindChainSplitResponse {
   // The current header height of this node
   uint64 tip_height = 3;
+  // The current geometric mean of the pow of the chain tip, or `None` if there is no chain
+  bytes accumulated_difficulty = 4;
```
this is meaning less, the peer can fill anything in here.  We already have the cl;aimed pow from the metadata ping/pong

hansieodendaal replied:
If the peer is honest, it have may increased since the ping-pong. If the peer is dishonest, they can lie about `tip_height` as well.

hansieodendaal replied:
If the peer is honest, it have may increased since the pin-pong. If the peer is dishonest, they can lie about `tip_height` as well.

SWvheerden replied:
well then lets remove that as well, all this does, is its asking the same info twice, if the peer already lied, why lie again. 

If the difficulty went up, it has no effect on the rest of the code. If this data is not required, let's remove it, it adds extra comms data, extra code complexity.

hansieodendaal replied:
Removed

SWvheerden commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -245,59 +248,72 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
+
+        match header_sync_status.clone() {
+            HeaderSyncStatus::InSyncOrAhead => {
+                // Our POW is less than the peer's POW, as verified before attempted header sync, therefor, if the peer
+                // did not supply any headers and we know we are behind, then we can ban the peer.
```
```suggestion
                // Our POW is less than the peer's POW, as verified before the attempted header sync, therefor, if the peer
                // did not supply any headers and they claimed we are behind, then we can ban the peer.
```

SWvheerden commented on base_layer/core/src/base_node/sync/header_sync/synchronizer.rs:
```diff
@@ -245,59 +248,72 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
+                        "Peer `{}` did not provide any headers although they have a better chain and more headers: their \
+                        difficulty: {}, our difficulty: {}. Peer will be banned.",
+                        sync_peer,
+                        peer_response.peer_accumulated_difficulty,
```
just use the sync peer, no need to add new chain metadatata we cannot verify it anyway.

SWvheerden reviewed (changes requested):
adding in a new field to the proto for the peer to also send is meaningless.
They already sent us one that they claim is higher than ours. All we have to do now is verify it, we should not ask for a new one.

Commit:
review comments

Commit:
Simplification

dependabot[bot] committed:
chore(deps): bump rustix from 0.37.23 to 0.37.26

Bumps [rustix](https://github.com/bytecodealliance/rustix) from 0.37.23 to 0.37.26.
- [Release notes](https://github.com/bytecodealliance/rustix/releases)
- [Commits](https://github.com/bytecodealliance/rustix/compare/v0.37.23...v0.37.26)

---
updated-dependencies:
- dependency-name: rustix
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
more simplification

Commit:
licence check

Commit:
Finding TARI-038

SWvheerden commented:
Fixes: #5816

SWvheerden commented on base_layer/core/src/base_node/sync/header_sync/error.rs:
```diff
@@ -56,6 +56,8 @@ pub enum BlockHeaderSyncError {
     ConnectivityError(#[from] ConnectivityError),
     #[error("Node is still not in sync. Sync will be retried with another peer if possible.")]
     NotInSync,
+    #[error("Internal data inconsistency found `{0}")]
```
I dont believe this is used anymore?

SWvheerden commented on base_layer/core/tests/helpers/sync.rs:
```diff
@@ -0,0 +1,261 @@
+            },
+        }
+        node.blockchain_db.write(txn).unwrap();
+        // Note: Something is funny here... the block is deleted but the block exists in the db
+        // match instruction {
+        //     WhatToDelete::Blocks => {
+        //         assert!(!node.blockchain_db.block_exists(*blocks[i].hash()).unwrap());
+        //     }
+        //     WhatToDelete::Headers => {}
+        // }
```
delete?

hansieodendaal replied:
No, I would like us to keep it here, would like to fix it soon, but it is not urgent right away.

SWvheerden replied:
Can you add a comment exactly saying it there as well, so we know why its commented out and why we left it in

SWvheerden commented on base_layer/core/tests/tests/header_sync.rs:
```diff
@@ -0,0 +1,373 @@
+        sync::create_and_add_some_blocks(&bob_node, &bob_blocks[1], 2, &consensus_manager, &key_manager, &[3; 2]).await;
+    assert_eq!(bob_node.blockchain_db.get_height().unwrap(), 4);
+
+    // Alice attempts header sync, on a higher chain with less POW, headers will be lagging with reorg steps
```
wait, This is unclear
If Bob's chain has more blocks, but less pow, Alice should not be lagging, or try syncing?

SWvheerden commented on base_layer/core/tests/tests/header_sync.rs:
```diff
@@ -0,0 +1,373 @@
+            assert_eq!(sync_result.headers_returned, 0);
+            assert_eq!(sync_result.peer_fork_hash_index, 3);
+            if let HeaderSyncStatus::InSyncOrAhead = sync_result.header_sync_status {
+                // Note: This behaviour is undetected! Bob cannot be banned.
```
```suggestion
                // Note: This behaviour is undetected! Bob cannot be banned here, and should be banned by block sync
```
This test and comment makes it seem like this is a bug and a viable attack. We should probably add a test for block sync where we test this.

hansieodendaal replied:
Agreed, it is a viable attack, and cannot be detected by header sync alone

SWvheerden reviewed (commented):
Looking good

Commit:
further review comments

Commit:
expand comment
